### PR TITLE
[SPARK-41441][SQL] Support Generate with no required child output to host outer references

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?
This PR supports `Generate` with no required child output to host outer references. It updates DecorrelateInnerQuery to handle Generate and adds a new optimization for Generate and OneRowRelation in rule OptimizeOneRowRelationSubquery.

### Why are the changes needed?
Currently, CheckAnalysis allows `Generate` with no required child output to host outer references. However,  DecorrelateInnerQuery does not handle this case correctly.  For example
```
select * from t, lateral (select explode(array(c1, c2)))
```
This will throw an internal error:
```
AssertionError: assertion failed: Correlated column is not allowed in Generate
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit tests and SQL query tests.

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala [hunk 1]
```diff
@@ -667,6 +667,18 @@ object DecorrelateInnerQuery extends PredicateHelper {
             val newJoin = j.copy(left = newLeft, right = newRight, condition = newCondition)
             (newJoin, newJoinCond, newOuterReferenceMap)
 
+          case g: Generate if g.requiredChildOutput.isEmpty =>
+            // Generate with non-empty required child output cannot host
+            // outer reference. It is blocked by CheckAnalysis.
+            val outerReferences = collectOuterReferences(g.expressions)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala:673]: is this always `Nil`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 2]
```diff
@@ -774,15 +775,28 @@ object OptimizeOneRowRelationSubquery extends Rule[LogicalPlan] {
...
+        case Project(projectList, _: OneRowRelation) =>
+          val newPList = stripOuterReferences(projectList)
+          Project(left.output ++ newPList, left)
+
+        case g @ Generate(generator, _, _, _, _, _: OneRowRelation) =>
+          val newGenerator = stripOuterReference(generator)
+          g.copy(generator = newGenerator, child = left)
+
+        case o =>
+          throw new IllegalStateException(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:790]: we should use `SparkThrowable.internalError`

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**